### ISSUE TITLE
Open Grant Application: Testground

### ISSUE THREAD

**Issue by brdji** [filecoin-project/devgrants#440]
# Open Grant Proposal: `Testground`

**Name of Project:**
Testground

**Proposal Category:**
 `devtools-libraries`

**Proposer:** 
Darko Brdareski - `brdji`

**(Optional) Technical Sponsor:** 
@BigLep 

**Do you agree to open source all work you do on behalf of this RFP and dual-license under MIT and APACHE2 licenses?:** Yes

# Project Description
Testground is a platform for testing, benchmarking, and simulating distributed and p2p systems at scale.
Due to lack of maintenance, it is no longer being used. Various issues, such as reporting false negatives, Docker container and network problems must be fixed.
Users have also requested new features and support for more languages.

## Value
1. Testground as a service (TaaS) is very useful for testing any new Filecoin features and projects. Being able to integrate it as a part of a CI pipeline will be of great benefit to many development teams.
2. Testground could be used to test and measure how other codebases (ie. not only Filecoin-related projects) would perform in a distributed environment

The main risk lies in not being able to fix issues that prevent the Testground from being used on all platforms. Since Testground depends heavily on Docker to work, any Docker issues will be immediately visible in the tool. Some of these issues are based on hardware architecture, and present on some platforms (e.g. Apple M1 laptops), but not present on others (e.g. Linux laptops). In such a scenario, the team will prepare a plan with possible workarounds and investigate new solutions.

## Deliverables
- Github CI Action to trigger specified Testground tests
- Test runner support for Go, JS, and Rust
- Additional features, agreed upon in the future with key stakeholders, once the first two milestones (see below) have been completed

## Development Roadmap

**Milestone(s)**
 
**Milestone 1**
   - Improve the TaaS cluster: Having discussed the current state of the cluster with PL stakeholders, the team believes the following improvements should be done before any other steps:
    1. Research and test the possibility of migrating the action cluster to a dedicated Kubernetes provider, such as Amazon EKS.
    2. Authentication for running jobs, to limit the load on the TaaS
    3. Clean up the state after running tests, so performance is not affected if a test crashes or takes too long
    4. Scale up the cluster by tweaking k8s parameters, taking request size and number into account
  - Update documentation and playbooks: Go over existing documentation and revise it to reflect the new state of the infrastructure
  - Issues triage: Investigate reported testground issues, prioritize them according to severity and impact, and fix them, starting with the top priority ones

The goal for this milestone is to fix the issues preventing the TaaS action cluster from being used as intended. We hope to determine the cost and possibility of migrating the cluster to a cloud-based Kubernetes provider; such a migration will allow for a wider array of configuration and usage options.
 
Est Milestone Date:
Q1 2022
 
**Milestone 2**
   - Migrate the TaaS cluster: If the tests performed in Milestone 1 show that a migration is possible, the cluster should be migrated to a dedicated Kubernetes provider as soon as possible. This will allow new features to be added, such as autoscaling and monitoring.
   - Add rust support: Add support for running Rust tests. This includes a builder, runner, as well as examples and documentation on writing and running Rust tests.
   - Fix and maintain JS support: According to several reports, the JS builder/runner is currently not working. Examine current JS support, fix it, and respond to any connected reports and issues.
   - Continue eliminating user-reported issues

 
The goal for this milestone is to migrate the cluster to a dedicated provider (if said migration was possible), and add new features and language support/
 
Est Milestone Date:
Q2 2022
 
**Milestone 3**
   - Continue fixing issues and maintain service integrity
   - Add new features: Browser test support
   - Add new features, starting with those most-requested by the community and key stakeholders
 
The goal for this milestone will be to implement features that have been selected by the community and stakeholders.
Since it is currently unclear what new features will be most impactful, the team working on the Testground will keep an ongoing
list of desired (and possible) features, from which key features will be selected.
 
Est Milestone Date:
Q3 2022

## Total Budget Requested

The project is time and material based, and the average hourly rate of the team is estimated at $75.
Our overall estimate is that each milestone will take between 1-2 months, based on core team feedback and the selected issues. As we move forward and get familiar with the project, more accurate estimates will be possible.

## Maintenance and Upgrade Plans

Once the outlined milestones have been completed, a new plan for long-term maintenance will be drafted by the team and stakeholders. The team sees the maintenance of this project as an something of great importance, both in accrued technical knowledge, and the value the Testground brings to other dev teams and the community.

# Team
Bloxico

## Key contributors
@brdji
@TheDivic
@TheMenko

## Team Member LinkedIn Profiles

https://www.linkedin.com/in/darko-brdareski-b7463b63/
https://www.linkedin.com/in/thedivic/
https://www.linkedin.com/in/nemanja-radenkovic-874090168/

## Relevant Experience
Our team is relatively new to the Filecoin ecosystem, composed of experienced developers with blockchain experience. We have worked on scalable infrastructure projects in the past, and are eager to work on, and improve, the Testground platform.
For the first phase, the team will be composed of 2 devops engineers, and 1 developer. This composition of roles reflects the needs of the Testground project in the early phase, where most of the tasks focus on improving the infrastructure of the service.
Once the main infrastructure workload has been cleared, we will restructure the team to include 3 developers and 1 devops engineer.

**Comment by BigLep** [comment]
Thanks @brdji so much for filing this.  This is exciting!

I am one of the engineering managers at Protocol Labs, and getting this kind of support for Testground has been identified as critical for our success in 2022.  I know Filecoin has needs here, but I'm actually most aware of some the needs within libp2p and IPFS for interoperability and performance tests.

Anyways, more to come soon here.  I just wanted to acknowledge that we've received this and will huddle internally to get back to you.

**Comment by BlocksOnAChain** [comment]
@brdji thanks for presenting and laying out your idea for testground.
I do agree that all the features you suggested would be great to have, but the main prerequisite that will unlock "feature building" would be to: get the infra to a stable state, update infra playbooks and documentation

**Comment by BigLep** [comment]
Thanks @BlocksOnAChain for providing feedback.  To make fully clear, before we get started on any of this work, we believe it's worth investing to see if we can get Testground running on managed infrastructure like Amazon EKS (this this relates to item 3 above of "Hardening of Kubernetes cluster".  

In case you hadn't seen this list, here is a dumping of things that the existing Testground team in at PL had called out as needing to happen for Testground: https://github.com/testground/testground/issues/1280 .  If/when we move forward with this grant, we can get all of this into GitHub issues on a project board in priority order.

A key question though is do you think you can take on the initial effort to shore up infrastructure?  PL would provide a DRI (directly responsible individual) to work with, and we'd use public communication channels.

**Comment by realChainLife** [comment]
Hi @brdji thank you for this proposal. We would like to fund the work outlined here, please email devgrants@fil.org to discuss next steps.

**Comment by BigLep** [comment]
@realChainLife : I believe @brdji and team are going to be making some updates based on the feedback we provided on where to focus first.

**Comment by brdji** [comment]
@realChainLife  Thank you! We are currently preparing updates to the roadmap, based on the feedback we have received. As soon as we are prepared, I will publish the updates to the proposal, and send an email to discuss the next steps.

**Comment by brdji** [comment]
I have updated the proposal to reflect received feedback. Milestone 1 has been revised, so the required infrastructure changes are done as early as possible in the project.

**Comment by BigLep** [comment]
@brdji : thanks for submitting.  This fell through the cracks on my end to review so we get this approved.  We'll get this looked at early this week.  Thanks!

**Comment by BigLep** [comment]
Also, I don't assume will get to this point in the next week, but in case you need dev contacts for this project in my absence, @nonsense @coryschwartz and @galargh will all be involved to some degree and aware that we're working to get this grant approved.

**Comment by BigLep** [comment]
For anyone watching this, I am back from being away and can respond more quickly.

More understanding of where we're at...
1. The Filecoin Foundation has "Approved" this grant (given the "Approved Label")
2. An email between @realChainLife , @brdji , and me was sent by the Filecoin Foundation concerning the Open Source Agreement.  

@brdji : I'm here to support getting this over the line.  Let me know if you need anything, either here, over email, or over a call.  

Thanks.

**Comment by BigLep** [comment]
@brdji : I know we're getting the grant specifics ironed out with Filecoin Foundation, but for coordinating on an engineering level, let's use IPFS Discord #testground-dev: https://discord.gg/dq8xd582aS
